### PR TITLE
feat: Add validating rules for Workspace name input

### PR DESCRIPTION
**Pull request by thaorell**
closes: #1304

- Add validating rules for Workspace name input. Disable Next button if the rule is violated
- Add tests

### CODE REVIEW

**Comment by thaorell** [comment]
https://github.com/user-attachments/assets/7118fb6f-a150-400e-ab9f-d9c651d96115

File: workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx [hunk 1]
```diff
@@ -143,7 +144,14 @@ const WorkspaceForm: React.FC = () => {
       setData('podConfig', undefined);
     }
   }, [filteredValuesData, filteredValuesLoaded, data.podConfig, setData]);
-
+  const onDisplayNameChange = useCallback(
+    (value: string) => {
+      const inputValue = value.trim();
```
Comment (christian-heusel) [inline workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx]: by calling `.trim()` on each input spaces are just being eaten as one types which is potentially confusing to the user 🤔

Comment (christian-heusel) [reply workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx]: Should we perhaps only call it once the user is "done typing" (idk what this could look like for the frontend, maybe whever the input in unfocused?). Also we could instead just replace spaces with dashes as the proposal initially had in mind.

**Review by christian-heusel — COMMENTED** [review]
Hey Charles, thanks a lot for coming up with this so quickly after our discussion on Slack! 🎉 

I found one thing that I stumbled across when testing this (both manually and agent based), see the review comment below.

<details>
<summary>Click to see agent test report</summary>

# Test Report: PR #1335 "feat: Add validating rules for Workspace name input"

**Issue:** [#1304 — Slugging for workspace name](https://github.com/kubeflow/notebooks/issues/1304)
**PR:** [#1335](https://github.com/kubeflow/notebooks/pull/1335) (author: Charles Thao)
**Commit tested:** `0cf73042` (branch `1304_slugging`)
**Tested by:** Christian Heusel (with Claude Code)
**Date:** 2026-08-18
**Scope:** Frontend only (`workspaces/frontend`)

---

## Summary

The PR adds **client-side validation** to the Workspace Name field in the create-workspace
wizard. Invalid names are highlighted in red with a specific error message, and the wizard's
**Next** button is disabled until the name is valid (and the other required fields are set).

All automated checks pass and the behaviour was confirmed live against the running Tilt
environment. **However, the PR implements a different UX than issue #1304 asks for** — see
the ⚠ "Scope divergence" section, which is the main item for reviewer judgement.

**Verdict:** Implementation is correct, well-tested, and merge-quality **for what it does
(validate-and-block)**. Whether that satisfies #1304 (which specifies *slugging* /
auto-normalisation) is a product decision that should be confirmed before closing the issue.

---

## What the PR actually implements

A new `validateName()` helper (`Form/helpers.tsx`) enforces, in order:

| Rule | Message | Notes |
|------|---------|-------|
| Non-empty | `Value is required` | after trimming |
| ≤ 63 chars | `Must be no more than 63 characters` | issue text says 253 (see below) |
| `^[a-z0-9.-]+$` | `Only lowercase alphanumeric characters, "-" or "." are allowed` | rejects uppercase, `_`, spaces, specials |
| starts alnum | `Must start with an alphanumeric character` | |
| ends alnum | `Must end with an alphanumeric character` | |

Wired in via `WorkspaceForm.tsx` (`onDisplayNameChange` → trims input, runs validator, stores
error) and `WorkspaceFormPropertiesSelection.tsx` (`ValidatedOptions.error` + helper text). The
error state also feeds the `canProceedToNextStep` gate for both the Properties and Summary steps.

---

## Automated test results

| Check | Command | Result |
|-------|---------|--------|
| TypeScript type-check | `npx tsc --noEmit` | ✅ pass (exit 0) |
| ESLint | `npm run test:lint` | ✅ pass (0 warnings) |
| Jest unit tests | `npm run test:jest` | ✅ 425 passed / 42 suites |
| Cypress (mocked) `createWorkspace.cy.ts` | `npm run test:cypress-ci` | ✅ 73 passing / 0 failing |

The PR adds 8 new Cypress cases under **"Workspace name validation"** (uppercase, invalid chars,
leading/trailing non-alnum, >63 chars, valid name, re-enable after correction, disable after
clear) — all green. Mock fixtures were also updated to use RFC1123-compliant names
(`my-first-jupyter-notebook`, etc.), which is why the surrounding suite needed touching.

---

## Live end-to-end verification (Tilt @ `https://localhost:8443`)

Walked the wizard to the **Properties** step (JupyterLab → default image → Tiny CPU →
attach `workspace-home-pvc`) and exercised the field. A screen recording was saved:
**`pr1335_workspace_name_validation.gif`**.

| Input | Observed | Correct? |
|-------|----------|----------|
| `MyWorkspace` (uppercase) | red field, "Only lowercase alphanumeric…", Next disabled | ✅ |
| `my_workspace!` (`_` + `!`) | red field, "Only lowercase alphanumeric…", Next disabled | ✅ |
| `-my-workspace` (leading `-`) | red field, "Must start with an alphanumeric character" | ✅ |
| `a`×64 (too long) | red field, "Must be no more than 63 characters", Next disabled | ✅ |
| `my-workspace.v1` (valid) | blue field, no error, **Next enabled** | ✅ |
| invalid name + volume attached | Next stays **disabled** (gate works with all other fields set) | ✅ |
| `my workspace` (space typed) | rendered as `myworkspace`, treated **valid** — space silently dropped | ⚠ see below |

---

## ⚠ Scope divergence — validation vs. slugging (main reviewer decision)

Issue #1304 is titled *"Slugging for workspace name"* and its acceptance criteria describe an
**auto-normalisation** flow: the user types a friendly display name (uppercase, spaces, specials
like `Bella's GPU Training Run`) and the app **generates** a valid Kubernetes name from it
(lowercase, spaces/underscores → dashes, strip invalid chars, trim leading `-`/`.`).

This PR takes the **opposite approach**: it *rejects* any non-conforming input and blocks
progress until the user types a name that is already RFC1123-compliant. There is **no slug
generation** and no separate display-name vs. resource-name concept.

This is a legitimate, simpler design — but it does **not** meet the issue's stated criteria as
written. This should be reconciled with the author/issue owner: either (a) accept validation as
the agreed scope and update/close #1304 accordingly, or (b) treat slugging as still-outstanding.

Concrete criteria from #1304 **not** met by this PR:
- Input does **not** permit uppercase, spaces, or special chars (`! @ # $ …`) — they're rejected, not slugged.
- Spaces are **not** converted to dashes; they are silently deleted (`my workspace` → `myworkspace`, not `my-workspace`).
- No underscore→dash conversion; `_` is rejected.
- No auto-lowercasing of user input.

---

## Minor / non-blocking observations

1. **Length limit 63 vs. 253.** The issue text says generated names may be up to 253 chars;
   the validator caps at 63. 63 is the stricter RFC1123 *label* limit and is safer for names
   that end up as DNS labels/pod names, so this won't produce invalid CRDs — but it's a
   deliberate deviation from the issue's number and worth a one-line confirmation.

2. **Space-swallowing UX.** Because `onDisplayNameChange` calls `.trim()` on every keystroke and
   feeds the trimmed value back as the controlled value, a space is stripped the instant it's
   typed — so interior spaces can never be entered at all (`my workspace` → `myworkspace`). The
   field silently changes what the user typed with no feedback. Minor, but a reviewer may want
   the value left as typed and simply flagged invalid, for clarity.

3. **Invalid value still shown in the Summary panel.** While a name is invalid, the right-hand
   Summary panel still echoes it (e.g. `Name: MyWorkspace`). Harmless since Next is blocked, but
   arguably shouldn't surface a value that can't be submitted.

4. **Frontend-only.** No backend/CRD-side validation is added or exercised here; this relies on
   the controller's existing name constraints as the source of truth.

---

## Recommendation

- ✅ **Code quality / tests / behaviour:** solid — approve on those grounds.
- ⚠ **Before merging as the resolution of #1304:** confirm with the author whether
  "validate-and-block" is the agreed replacement for the "slugging" behaviour the issue
  specifies. If slugging is still wanted, this PR is a partial step and #1304 should stay open.
- 🔧 **Optional polish:** reconsider the trim-on-every-keystroke behaviour (obscures typed
  spaces) and the 63 vs. 253 length choice.

</details>

Since the PR does not implement the feature as stated in the AC we shouldn't close the feature even once the PR is merged, however it is a really valuable contribution to spare the Data Scientist persona from any confusion in the context of the Beta release.

 I will instead bump this for discussion to the 2.0.0/GA target once we merged this PR 🤗

File: workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx (hunk 1 above)
Comment (thaorell) [reply workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx]: Removing the `.trim()` in favor of displaying the error to the user here

Comment (christian-heusel) [reply workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx]: Resolved with the latest changes ✅ Thanks a lot!

**Review by christian-heusel — APPROVED** [review]

File: workspaces/frontend/src/app/pages/Workspaces/Form/properties/WorkspaceFormPropertiesSelection.tsx [hunk 2]
```diff
@@ -76,13 +86,15 @@ const WorkspaceFormPropertiesSelection: React.FunctionComponent<
             isRequired
             fieldId="workspace-name"
             className="pf-u-width-520"
+            helperTextNode={workspaceNameError}
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/properties/WorkspaceFormPropertiesSelection.tsx]: The `helperTextNode` prop receives a raw error string, but the established pattern in this codebase wraps error messages in `<HelperText><HelperTextItem variant="error">{message}</HelperTextItem></HelperText>` (see `MountPathField.tsx`). Passing a raw string may render without proper PatternFly error styling. Consider wrapping: `helperTextNode={workspaceNameError ? <HelperText><HelperTextItem variant="error">{workspaceNameError}</HelperTextItem></HelperText> : null}`.

File: workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx [hunk 3]
```diff
@@ -105,7 +106,7 @@ const WorkspaceForm: React.FC = () => {
 
   // Store original values for edit mode diff view
   const [originalData, setOriginalData] = useState<WorkspaceFormData | undefined>(undefined);
-
+  const [displayNameError, setDisplayNameError] = useState<string | null>(null);
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx]: Nit: the blank line between `originalData` state and the ref declarations was removed, making `displayNameError` visually run together with `originalData` and the comment above it. Adding back a blank line before `const [displayNameError, ...]` would maintain readability and separate the logical groupings (original data tracking vs. validation state).

File: workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx [hunk 4]
```diff
@@ -143,7 +144,13 @@ const WorkspaceForm: React.FC = () => {
       setData('podConfig', undefined);
     }
   }, [filteredValuesData, filteredValuesLoaded, data.podConfig, setData]);
-
+  const onDisplayNameChange = useCallback(
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/WorkspaceForm.tsx:149]: Nit: `onDisplayNameChange` callback and the preceding `useEffect` closing run together without a blank line separator. Adding blank lines between logical blocks (effect hooks, callback definitions) improves scanability, consistent with the rest of this file.

File: workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx [hunk 5]
```diff
@@ -274,3 +274,26 @@ export const buildPVCSelectOptions = (
 
   return options;
 };
+
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx:278]: The `validateName` function is a pure utility with multiple branches — it's an ideal candidate for a dedicated unit test file (e.g., `__tests__/helpers.spec.ts`), similar to how `validateMountPath` and other helpers are structured. The Cypress tests cover integration behavior, but unit tests would catch edge cases more directly and run faster. Consider adding unit tests covering each validation branch (empty, too long, invalid chars, invalid start/end, valid input).

File: workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx [hunk 6]
```diff
@@ -274,3 +274,26 @@ export const buildPVCSelectOptions = (
...
+export const validateName = (name: string): string | null => {
+  if (!name) {
+    return 'Value is required';
+  }
+
+  if (name.length > 63) {
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx]: The max-length threshold `63` is a magic number. Consider extracting it into a named constant (e.g., `const MAX_WORKSPACE_NAME_LENGTH = 63;`) at the top of the file or in `const.ts`, consistent with how `MOUNT_PATH_MAX_LENGTH` is already defined a few lines above.

**Review by paulovmr — CHANGES_REQUESTED** [review]

File: workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx [hunk 7]
```diff
@@ -274,3 +275,26 @@ export const buildPVCSelectOptions = (
...
   return options;
 };
+
+export const validateName = (name: string): string | null => {
+  if (!name) {
+    return 'Value is required';
+  }
+
+  if (name.length > MAX_WORKSPACE_NAME_LENGTH) {
+    return 'Must be no more than 63 characters';
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/helpers.tsx]: The condition correctly uses `MAX_WORKSPACE_NAME_LENGTH`, but the error message hardcodes "63": `'Must be no more than 63 characters'`. If the constant changes, the message won't update. Consider using a template literal: `` `Must be no more than ${MAX_WORKSPACE_NAME_LENGTH} characters` ``.

File: workspaces/frontend/src/app/pages/Workspaces/Form/utils/__tests__/helpers.spec.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  });
+
+  it('should end with an alphanumeric character', () => {
+    const name = 'a-';
+
+    const result = validateName(name);
+
+    expect(result).toEqual('Must end with an alphanumeric character');
+  });
+  it('should be valid', () => {
```
Comment (paulovmr) [inline workspaces/frontend/src/app/pages/Workspaces/Form/utils/__tests__/helpers.spec.ts:44]: Two suggestions: (1) Add a test case for empty string input — the `'Value is required'` branch is untested. (2) Nit: missing blank line before the last `it('should be valid', ...)` block (line 35) — the other test cases each have a blank line separator.

**Review by paulovmr — CHANGES_REQUESTED** [review]

**Review by paulovmr — APPROVED** [review]
/lgtm
/approve
